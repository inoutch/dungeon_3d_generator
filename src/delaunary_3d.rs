use nalgebra::{Matrix4, Vector3};
use std::collections::HashSet;
use std::hash::{Hash, Hasher};

const ACCURACY: f32 = 1000.0;

///
/// Reference: https://github.com/vazgriz/DungeonGenerator/blob/master/Assets/Scripts3D/Delaunay3D.cs
///
#[derive(Debug, Clone)]
pub struct Vertex {
    pub position: Vector3<f32>,
}

impl PartialEq for Vertex {
    fn eq(&self, other: &Self) -> bool {
        (
            (self.position.x * ACCURACY) as i64,
            (self.position.y * ACCURACY) as i64,
            (self.position.z * ACCURACY) as i64,
        ) == (
            (other.position.x * ACCURACY) as i64,
            (other.position.y * ACCURACY) as i64,
            (other.position.z * ACCURACY) as i64,
        )
    }
}

impl Eq for Vertex {}

impl Hash for Vertex {
    fn hash<H: Hasher>(&self, state: &mut H) {
        (
            (self.position.x * ACCURACY) as i64,
            (self.position.y * ACCURACY) as i64,
            (self.position.z * ACCURACY) as i64,
        )
            .hash(state);
    }
}

#[derive(Debug, Clone)]
pub struct Tetrahedron {
    pub a: Vertex,
    pub b: Vertex,
    pub c: Vertex,
    pub d: Vertex,
    pub is_bad: bool,
    circumcenter: Vector3<f32>,
    circumradius_squared: f32,
}

impl Tetrahedron {
    pub fn new(a: Vertex, b: Vertex, c: Vertex, d: Vertex) -> Self {
        let mut tetra = Tetrahedron {
            a,
            b,
            c,
            d,
            is_bad: false,
            circumcenter: Vector3::zeros(),
            circumradius_squared: 0.0,
        };
        tetra.calculate_circumsphere();
        tetra
    }

    fn calculate_circumsphere(&mut self) {
        // Matrix determinant calculation for circumcenter and circumradius
        let a_matrix = Matrix4::new(
            self.a.position.x,
            self.b.position.x,
            self.c.position.x,
            self.d.position.x,
            self.a.position.y,
            self.b.position.y,
            self.c.position.y,
            self.d.position.y,
            self.a.position.z,
            self.b.position.z,
            self.c.position.z,
            self.d.position.z,
            1.0,
            1.0,
            1.0,
            1.0,
        );
        let det_a = a_matrix.determinant();

        let pos_sqr_a = self.a.position.norm_squared();
        let pos_sqr_b = self.b.position.norm_squared();
        let pos_sqr_c = self.c.position.norm_squared();
        let pos_sqr_d = self.d.position.norm_squared();

        let dx_matrix = Matrix4::new(
            pos_sqr_a,
            pos_sqr_b,
            pos_sqr_c,
            pos_sqr_d,
            self.a.position.y,
            self.b.position.y,
            self.c.position.y,
            self.d.position.y,
            self.a.position.z,
            self.b.position.z,
            self.c.position.z,
            self.d.position.z,
            1.0,
            1.0,
            1.0,
            1.0,
        );
        let dx = dx_matrix.determinant();

        let dy_matrix = Matrix4::new(
            pos_sqr_a,
            pos_sqr_b,
            pos_sqr_c,
            pos_sqr_d,
            self.a.position.x,
            self.b.position.x,
            self.c.position.x,
            self.d.position.x,
            self.a.position.z,
            self.b.position.z,
            self.c.position.z,
            self.d.position.z,
            1.0,
            1.0,
            1.0,
            1.0,
        );
        let dy = -dy_matrix.determinant();

        let dz_matrix = Matrix4::new(
            pos_sqr_a,
            pos_sqr_b,
            pos_sqr_c,
            pos_sqr_d,
            self.a.position.x,
            self.b.position.x,
            self.c.position.x,
            self.d.position.x,
            self.a.position.y,
            self.b.position.y,
            self.c.position.y,
            self.d.position.y,
            1.0,
            1.0,
            1.0,
            1.0,
        );
        let dz = dz_matrix.determinant();

        let c_matrix = Matrix4::new(
            pos_sqr_a,
            pos_sqr_b,
            pos_sqr_c,
            pos_sqr_d,
            self.a.position.x,
            self.b.position.x,
            self.c.position.x,
            self.d.position.x,
            self.a.position.y,
            self.b.position.y,
            self.c.position.y,
            self.d.position.y,
            self.a.position.z,
            self.b.position.z,
            self.c.position.z,
            self.d.position.z,
        );
        let det_c = c_matrix.determinant();

        self.circumcenter =
            Vector3::new(dx / (2.0 * det_a), dy / (2.0 * det_a), dz / (2.0 * det_a));
        self.circumradius_squared =
            (dx * dx + dy * dy + dz * dz - 4.0 * det_a * det_c) / (4.0 * det_a * det_a);
    }

    pub fn circum_circle_contains(&self, v: &Vector3<f32>) -> bool {
        let dist = v - self.circumcenter;
        dist.norm_squared() <= self.circumradius_squared
    }

    pub fn contains_vertex(&self, v: &Vertex) -> bool {
        v == &self.a || v == &self.b || v == &self.c || v == &self.d
    }
}

#[derive(Clone, Debug, Eq)]
pub struct Triangle {
    pub u: Vertex,
    pub v: Vertex,
    pub w: Vertex,
    pub is_bad: bool,
}

impl Hash for Triangle {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.u.hash(state);
        self.v.hash(state);
        self.w.hash(state);
    }
}

impl Triangle {
    pub fn new(u: Vertex, v: Vertex, w: Vertex) -> Self {
        Self {
            u,
            v,
            w,
            is_bad: false,
        }
    }
}

impl PartialEq for Triangle {
    fn eq(&self, other: &Self) -> bool {
        (self.u == other.u || self.u == other.v || self.u == other.w)
            && (self.v == other.u || self.v == other.v || self.v == other.w)
            && (self.w == other.u || self.w == other.v || self.w == other.w)
    }
}

#[derive(Debug, Clone, Eq)]
pub struct Edge {
    pub u: Vertex,
    pub v: Vertex,
    pub is_bad: bool,
}

impl Hash for Edge {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.u.hash(state);
        self.v.hash(state);
    }
}

impl Edge {
    pub fn new(u: Vertex, v: Vertex) -> Self {
        Self {
            u,
            v,
            is_bad: false,
        }
    }
}

impl PartialEq for Edge {
    fn eq(&self, other: &Self) -> bool {
        (self.u == other.u || self.v == other.u) && (self.u == other.v || self.v == other.v)
    }
}

#[derive(Clone, Debug)]
pub struct Delaunay3D {
    pub vertices: Vec<Vertex>,
    pub edges: Vec<Edge>,
    pub triangles: Vec<Triangle>,
    pub tetrahedra: Vec<Tetrahedron>,
}

impl Delaunay3D {
    pub fn new(vertices: Vec<Vector3<f32>>) -> Self {
        let mut ret = Self {
            vertices: vertices
                .into_iter()
                .map(|v| Vertex { position: v })
                .collect(),
            edges: Vec::new(),
            triangles: Vec::new(),
            tetrahedra: Vec::new(),
        };
        ret.triangulate();
        ret
    }

    fn triangulate(&mut self) {
        let mut min_x = self.vertices[0].position.x;
        let mut min_y = self.vertices[0].position.y;
        let mut min_z = self.vertices[0].position.z;
        let mut max_x = min_x;
        let mut max_y = min_y;
        let mut max_z = min_z;

        for vertex in self.vertices.iter() {
            if vertex.position.x < min_x {
                min_x = vertex.position.x;
            }
            if vertex.position.x > max_x {
                max_x = vertex.position.x;
            }
            if vertex.position.y < min_y {
                min_y = vertex.position.y;
            }
            if vertex.position.y > max_y {
                max_y = vertex.position.y;
            }
            if vertex.position.z < min_z {
                min_z = vertex.position.z;
            }
            if vertex.position.z > max_z {
                max_z = vertex.position.z;
            }
        }

        let dx = max_x - min_x;
        let dy = max_y - min_y;
        let dz = max_z - min_z;
        let delta_max = dx.max(dy.max(dz)) * 2.0;

        let p1 = Vertex {
            position: Vector3::new(min_x - 1.0, min_y - 1.0, min_z - 1.0),
        };
        let p2 = Vertex {
            position: Vector3::new(max_x + delta_max, min_y - 1.0, min_z - 1.0),
        };
        let p3 = Vertex {
            position: Vector3::new(min_x - 1.0, max_y + delta_max, min_z - 1.0),
        };
        let p4 = Vertex {
            position: Vector3::new(min_x - 1.0, min_y - 1.0, max_z + delta_max),
        };

        self.tetrahedra.push(Tetrahedron::new(
            p1.clone(),
            p2.clone(),
            p3.clone(),
            p4.clone(),
        ));

        for vertex in self.vertices.iter() {
            let mut triangles = Vec::new();
            for tetrahedron in self.tetrahedra.iter_mut() {
                if tetrahedron.circum_circle_contains(&vertex.position) {
                    tetrahedron.is_bad = true;
                    triangles.push(Triangle::new(
                        tetrahedron.a.clone(),
                        tetrahedron.b.clone(),
                        tetrahedron.c.clone(),
                    ));
                    triangles.push(Triangle::new(
                        tetrahedron.a.clone(),
                        tetrahedron.b.clone(),
                        tetrahedron.d.clone(),
                    ));
                    triangles.push(Triangle::new(
                        tetrahedron.a.clone(),
                        tetrahedron.c.clone(),
                        tetrahedron.d.clone(),
                    ));
                    triangles.push(Triangle::new(
                        tetrahedron.b.clone(),
                        tetrahedron.c.clone(),
                        tetrahedron.d.clone(),
                    ));
                }
            }

            for i in 0..triangles.len() {
                for j in (i + 1)..triangles.len() {
                    if triangles[i] == triangles[j] {
                        triangles[i].is_bad = true;
                        triangles[j].is_bad = true;
                    }
                }
            }

            self.tetrahedra.retain(|tetrahedron| !tetrahedron.is_bad);
            triangles.retain(|triangle| !triangle.is_bad);

            for triangle in triangles {
                self.tetrahedra.push(Tetrahedron::new(
                    triangle.u,
                    triangle.v,
                    triangle.w,
                    vertex.clone(),
                ));
            }
        }

        self.tetrahedra.retain(|tetrahedron| {
            !tetrahedron.contains_vertex(&p1)
                && !tetrahedron.contains_vertex(&p2)
                && !tetrahedron.contains_vertex(&p3)
                && !tetrahedron.contains_vertex(&p4)
        });

        let mut triangle_set = HashSet::new();
        let mut edge_set = HashSet::new();

        for tetrahedron in self.tetrahedra.iter() {
            let abc = Triangle::new(
                tetrahedron.a.clone(),
                tetrahedron.b.clone(),
                tetrahedron.c.clone(),
            );
            let abd = Triangle::new(
                tetrahedron.a.clone(),
                tetrahedron.b.clone(),
                tetrahedron.d.clone(),
            );
            let acd = Triangle::new(
                tetrahedron.a.clone(),
                tetrahedron.c.clone(),
                tetrahedron.d.clone(),
            );
            let bcd = Triangle::new(
                tetrahedron.b.clone(),
                tetrahedron.c.clone(),
                tetrahedron.d.clone(),
            );

            if triangle_set.insert(abc.clone()) {
                self.triangles.push(abc);
            }
            if triangle_set.insert(abd.clone()) {
                self.triangles.push(abd);
            }
            if triangle_set.insert(acd.clone()) {
                self.triangles.push(acd);
            }
            if triangle_set.insert(bcd.clone()) {
                self.triangles.push(bcd);
            }

            let ab = Edge::new(tetrahedron.a.clone(), tetrahedron.b.clone());
            let bc = Edge::new(tetrahedron.b.clone(), tetrahedron.c.clone());
            let ca = Edge::new(tetrahedron.c.clone(), tetrahedron.a.clone());
            let da = Edge::new(tetrahedron.d.clone(), tetrahedron.a.clone());
            let db = Edge::new(tetrahedron.d.clone(), tetrahedron.b.clone());
            let dc = Edge::new(tetrahedron.d.clone(), tetrahedron.c.clone());

            if edge_set.insert(ab.clone()) {
                self.edges.push(ab);
            }
            if edge_set.insert(bc.clone()) {
                self.edges.push(bc);
            }
            if edge_set.insert(ca.clone()) {
                self.edges.push(ca);
            }
            if edge_set.insert(da.clone()) {
                self.edges.push(da);
            }
            if edge_set.insert(db.clone()) {
                self.edges.push(db);
            }
            if edge_set.insert(dc.clone()) {
                self.edges.push(dc);
            }
        }
    }
}
