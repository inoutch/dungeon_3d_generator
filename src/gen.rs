use crate::delaunary_3d::{Delaunay3D, Edge};
use nalgebra::Vector3;
use pathfinding::prelude::kruskal;
use rand::{Rng, SeedableRng};
use std::collections::BTreeMap;
use std::ops::RangeInclusive;
use std::rc::Rc;

pub struct Dungeon3DGeneratorConfig {
    pub width: u32,        // Width of entire dungeon (x-axis)
    pub height: u32,       // Height of entire dungeon (y-axis)
    pub depth: u32,        // Depth of entire dungeon (z-axis)
    pub seed: Option<u64>, // Seed value for random dungeon generation
    pub room_hierarchy: u32,
    pub room_width_range: RangeInclusive<u32>,
    pub room_height_range: RangeInclusive<u32>,
    pub room_depth_range: RangeInclusive<u32>,
    pub room_margin: u32,
}

impl Default for Dungeon3DGeneratorConfig {
    fn default() -> Self {
        Dungeon3DGeneratorConfig {
            width: 32,
            height: 16,
            depth: 32,
            seed: None,
            room_hierarchy: 3,
            room_width_range: 5..=10,
            room_height_range: 2..=3,
            room_depth_range: 5..=10,
            room_margin: 2,
        }
    }
}

#[derive(Debug)]
pub struct Room {
    pub id: RoomId,
    pub width: u32,
    pub height: u32,
    pub depth: u32,
    pub origin: (u32, u32, u32),
    pub center_offset: (f32, f32, f32),
}

impl Room {
    pub fn new(id: RoomId, width: u32, height: u32, depth: u32, origin: (u32, u32, u32)) -> Self {
        Room {
            id,
            width,
            height,
            depth,
            origin,
            center_offset: (width as f32 / 2.0, height as f32 / 2.0, depth as f32 / 2.0),
        }
    }

    pub fn center(&self) -> (f32, f32, f32) {
        (
            self.center_offset.0 + self.origin.0 as f32,
            self.center_offset.1 + self.origin.1 as f32,
            self.center_offset.2 + self.origin.2 as f32,
        )
    }

    pub fn end(&self) -> (u32, u32, u32) {
        (
            self.origin.0 + self.width,
            self.origin.1 + self.height,
            self.origin.2 + self.depth,
        )
    }

    pub fn is_contract(&self, other: &Room, margin: u32) -> bool {
        let self_end = self.end();
        let self_end = (
            self_end.0 + margin,
            self_end.1 + margin,
            self_end.2 + margin,
        );
        let other_end = other.end();
        let other_end = (
            other_end.0 + margin,
            other_end.1 + margin,
            other_end.2 + margin,
        );
        self.origin.0 <= other_end.0
            && other.origin.0 <= self_end.0
            && self.origin.1 <= other_end.1
            && other.origin.1 <= self_end.1
            && self.origin.2 <= other_end.2
            && other.origin.2 <= self_end.2
    }
}

#[derive(Ord, PartialOrd, PartialEq, Eq, Hash, Copy, Clone, Debug)]
pub struct RoomId(u64);

impl RoomId {
    pub fn first() -> Self {
        RoomId(1)
    }

    pub fn gen_id(&mut self) -> Self {
        let ret = *self;
        self.0 += 1;
        ret
    }
}

#[derive(Debug)]
pub struct RoomConnection {
    pub room0_id: RoomId,
    pub room1_id: RoomId,
    pub squared_length: f32,
}

#[derive(Debug)]
pub struct Dungeon3DGeneratorResult {
    pub rooms: BTreeMap<RoomId, Room>,
    pub room_connections: Vec<Rc<RoomConnection>>,
    pub edges: Vec<Edge>,
}

#[derive(Debug)]
pub enum Dungeon3DGeneratorError {
    NarrowWidthOrRoomWidthTooLarge,
    NarrowDepthOrRoomDepthTooLarge,
    NarrowHeightOrRoomHierarchyTooSmall,
}

pub fn generate_dungeon_3d(
    config: Dungeon3DGeneratorConfig,
) -> Result<Dungeon3DGeneratorResult, Dungeon3DGeneratorError> {
    // validate
    let w_divisions_max = config.width / (config.room_width_range.start() + config.room_margin);
    let w_divisions_min = config.width / (config.room_width_range.end() + config.room_margin);
    if w_divisions_min == 0 {
        return Err(Dungeon3DGeneratorError::NarrowWidthOrRoomWidthTooLarge);
    }
    let d_divisions_max = config.width / (config.room_depth_range.start() + config.room_margin);
    let d_divisions_min = config.width / (config.room_depth_range.end() + config.room_margin);
    if d_divisions_min == 0 {
        return Err(Dungeon3DGeneratorError::NarrowDepthOrRoomDepthTooLarge);
    }
    if config.room_hierarchy * (config.room_height_range.start() + config.room_margin)
        > config.height
    {
        return Err(Dungeon3DGeneratorError::NarrowHeightOrRoomHierarchyTooSmall);
    }

    let mut rng: rand::rngs::StdRng = config
        .seed
        .map(SeedableRng::seed_from_u64)
        .unwrap_or_else(rand::rngs::StdRng::from_entropy);

    let mut room_id = RoomId::first();
    let mut rooms = BTreeMap::new();
    let mut room_ids = Vec::new();
    let h_block_size = config.height / config.room_hierarchy;
    for ry in 0..config.room_hierarchy {
        let w_divisions = rng.gen_range(1..=w_divisions_max);
        let w_block_size = config.width / w_divisions;
        for rx in 0..w_divisions {
            let d_divisions = rng.gen_range(1..=d_divisions_max);
            let d_block_size = config.depth / d_divisions;
            for rz in 0..d_divisions {
                let room_width = rng.gen_range(
                    *config.room_width_range.start()
                        ..=(w_block_size - config.room_margin).min(*config.room_width_range.end()),
                );
                let room_height = rng.gen_range(
                    *config.room_height_range.start()
                        ..=(h_block_size - config.room_margin).min(*config.room_height_range.end()),
                );
                let room_depth = rng.gen_range(
                    *config.room_depth_range.start()
                        ..=(d_block_size - config.room_margin).min(*config.room_depth_range.end()),
                );
                let (origin_x, origin_y, origin_z) =
                    (rx * w_block_size, ry * h_block_size, rz * d_block_size);
                let room_origin = (
                    origin_x + rng.gen_range(0..=(w_block_size - room_width - config.room_margin)),
                    origin_y + rng.gen_range(0..=(h_block_size - room_height - config.room_margin)),
                    origin_z + rng.gen_range(0..=(d_block_size - room_depth - config.room_margin)),
                );
                let new_room_id = room_id.gen_id();
                room_ids.push(new_room_id);
                rooms.insert(
                    new_room_id,
                    Room::new(
                        new_room_id,
                        room_width,
                        room_height,
                        room_depth,
                        room_origin,
                    ),
                );
            }
        }
    }

    let center = (
        config.width as f32 / 2.0f32,
        config.height as f32 / 2.0,
        config.depth as f32 / 2.0f32,
    );
    let mut sorted_rooms = rooms
        .values()
        .map(|room| {
            let room_center = room.center();
            let diff = (
                room_center.0 - center.0,
                room_center.1 - center.1,
                room_center.2 - center.2,
            );
            let squared_length = diff.0 * diff.0 + diff.1 * diff.1 + diff.2 * diff.2;
            (room.id, squared_length)
        })
        .collect::<Vec<_>>();
    sorted_rooms.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap());

    let mut room_connections = Vec::new();
    let mut room_connection_map: BTreeMap<RoomId, BTreeMap<RoomId, Rc<RoomConnection>>> =
        BTreeMap::new();
    for room_index in 0..room_ids.len() {
        let current_room_id = room_ids[room_index];
        let current_room = rooms.get(&current_room_id).unwrap();
        let current_room_center = current_room.center();
        for target_room_id in &room_ids[(room_index + 1)..rooms.len()] {
            let target_room = rooms.get(target_room_id).unwrap();
            let target_room_center = target_room.center();
            let diff = (
                current_room_center.0 - target_room_center.0,
                current_room_center.1 - target_room_center.1,
                current_room_center.2 - target_room_center.2,
            );
            let squared_length = diff.0 * diff.0 + diff.1 * diff.1 + diff.2 * diff.2;
            let room_connection = Rc::new(RoomConnection {
                room0_id: current_room.id,
                room1_id: target_room.id,
                squared_length,
            });
            room_connections.push(room_connection.clone());
            room_connection_map
                .entry(target_room.id)
                .or_default()
                .insert(current_room.id, room_connection.clone());
            room_connection_map
                .entry(current_room.id)
                .or_default()
                .insert(target_room.id, room_connection.clone());
        }
    }

    let delaunay = Delaunay3D::new(
        rooms
            .values()
            .map(|room| {
                let center = room.center();
                Vector3::new(center.0, center.1, center.2)
            })
            .collect(),
    );

    for (_, room) in rooms.iter_mut() {
        let mut room_origin = room.origin;
    }

    // Create mst of room neighbors
    let weighted_edges = room_connections
        .iter()
        .map(|room_connection| {
            (
                room_connection.room0_id,
                room_connection.room1_id,
                room_connection.squared_length as u64,
            )
        })
        .collect::<Vec<_>>();
    let room_connections = kruskal(&weighted_edges)
        .map(|(room0_id, room1_id, _)| {
            Rc::clone(
                room_connection_map
                    .get(room0_id)
                    .unwrap()
                    .get(room1_id)
                    .unwrap(),
            )
        })
        .collect::<Vec<_>>();

    Ok(Dungeon3DGeneratorResult {
        rooms,
        room_connections,
        edges: delaunay.edges,
    })
}

#[cfg(test)]
mod tests {
    use crate::gen::{generate_dungeon_3d, Dungeon3DGeneratorConfig};

    #[test]
    fn test_default_generate() {
        let result = generate_dungeon_3d(Dungeon3DGeneratorConfig {
            seed: Some(0),
            ..Default::default()
        })
        .unwrap();
        insta::assert_debug_snapshot!(result);
    }
}
