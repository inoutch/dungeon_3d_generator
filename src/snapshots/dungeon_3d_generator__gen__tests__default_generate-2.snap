---
source: src/gen.rs
expression: result.rooms
---
{
    RoomId(
        1,
    ): Room {
        id: RoomId(
            1,
        ),
        width: 6,
        height: 2,
        depth: 6,
        origin: (
            0,
            0,
            0,
        ),
        center_offset: (
            3.0,
            1.0,
            3.0,
        ),
    },
    RoomId(
        2,
    ): Room {
        id: RoomId(
            2,
        ),
        width: 5,
        height: 2,
        depth: 5,
        origin: (
            0,
            0,
            10,
        ),
        center_offset: (
            2.5,
            1.0,
            2.5,
        ),
    },
    RoomId(
        3,
    ): Room {
        id: RoomId(
            3,
        ),
        width: 5,
        height: 2,
        depth: 6,
        origin: (
            0,
            0,
            20,
        ),
        center_offset: (
            2.5,
            1.0,
            3.0,
        ),
    },
    RoomId(
        4,
    ): Room {
        id: RoomId(
            4,
        ),
        width: 6,
        height: 2,
        depth: 9,
        origin: (
            10,
            0,
            0,
        ),
        center_offset: (
            3.0,
            1.0,
            4.5,
        ),
    },
    RoomId(
        5,
    ): Room {
        id: RoomId(
            5,
        ),
        width: 6,
        height: 2,
        depth: 7,
        origin: (
            10,
            0,
            17,
        ),
        center_offset: (
            3.0,
            1.0,
            3.5,
        ),
    },
    RoomId(
        6,
    ): Room {
        id: RoomId(
            6,
        ),
        width: 6,
        height: 2,
        depth: 5,
        origin: (
            20,
            0,
            5,
        ),
        center_offset: (
            3.0,
            1.0,
            2.5,
        ),
    },
    RoomId(
        7,
    ): Room {
        id: RoomId(
            7,
        ),
        width: 5,
        height: 2,
        depth: 10,
        origin: (
            21,
            0,
            18,
        ),
        center_offset: (
            2.5,
            1.0,
            5.0,
        ),
    },
    RoomId(
        8,
    ): Room {
        id: RoomId(
            8,
        ),
        width: 6,
        height: 2,
        depth: 10,
        origin: (
            4,
            3,
            5,
        ),
        center_offset: (
            3.0,
            1.0,
            5.0,
        ),
    },
    RoomId(
        9,
    ): Room {
        id: RoomId(
            9,
        ),
        width: 5,
        height: 2,
        depth: 10,
        origin: (
            0,
            6,
            0,
        ),
        center_offset: (
            2.5,
            1.0,
            5.0,
        ),
    },
    RoomId(
        10,
    ): Room {
        id: RoomId(
            10,
        ),
        width: 6,
        height: 2,
        depth: 6,
        origin: (
            6,
            6,
            21,
        ),
        center_offset: (
            3.0,
            1.0,
            3.0,
        ),
    },
    RoomId(
        11,
    ): Room {
        id: RoomId(
            11,
        ),
        width: 8,
        height: 2,
        depth: 7,
        origin: (
            16,
            6,
            1,
        ),
        center_offset: (
            4.0,
            1.0,
            3.5,
        ),
    },
    RoomId(
        12,
    ): Room {
        id: RoomId(
            12,
        ),
        width: 9,
        height: 2,
        depth: 5,
        origin: (
            16,
            6,
            17,
        ),
        center_offset: (
            4.5,
            1.0,
            2.5,
        ),
    },
}
